examples/escrow/programs/escrow/src/lib.rs

```diff
@@ -0,0 +1,238 @@
+        ctx.accounts
+            .escrow_account
+            .initializer_deposit_token_account = *ctx
+            .accounts
+            .initializer_deposit_token_account
+            .to_account_info()
+            .key;
+        ctx.accounts
+            .escrow_account
+            .initializer_receive_token_account = *ctx
+            .accounts
+            .initializer_receive_token_account
+            .to_account_info()
+            .key;
+        ctx.accounts.escrow_account.initializer_amount = initializer_amount;
+        ctx.accounts.escrow_account.taker_amount = taker_amount;
+
+        let (pda, _bump_seed) = Pubkey::find_program_address(&[b"escrow"], ctx.program_id);
+        let set_authority_accounts = SetAuthority {
+            account_or_mint: ctx
+                .accounts
+                .initializer_deposit_token_account
+                .to_account_info()
+                .clone(),
+            current_authority: ctx.accounts.initializer.clone(),
+        };
+        let cpi_context =
+            CpiContext::new(ctx.accounts.token_program.clone(), set_authority_accounts);
+
+        token::set_authority(cpi_context, AuthorityType::AccountOwner, Some(pda))?;
+        Ok(())
+    }
+
+    pub fn cancel_escrow(ctx: Context<CancelEscrow>) -> ProgramResult {
+        let (_pda, bump_seed) = Pubkey::find_program_address(&[b"escrow"], ctx.program_id);
+        let set_authority_accounts = SetAuthority {
+            account_or_mint: ctx
+                .accounts
+                .pda_deposit_token_account
+                .to_account_info()
+                .clone(),
+            current_authority: ctx.accounts.pda_account.clone(),
+        };
+
+        let seeds = &[&b"escrow"[..], &[bump_seed]];
+        let signer = &[&seeds[..]];
+
+        let cpi_context = CpiContext::new_with_signer(
+            ctx.accounts.token_program.clone(),
+            set_authority_accounts,
+            signer,
+        );
+
+        token::set_authority(
+            cpi_context,
+            AuthorityType::AccountOwner,
+            Some(ctx.accounts.escrow_account.initializer_key),
+        )?;
+
+        Ok(())
+    }
+
+    pub fn exchange(ctx: Context<Exchange>) -> ProgramResult {
+        // Transferring from initializer to taker
+        let (_pda, bump_seed) = Pubkey::find_program_address(&[b"escrow"], ctx.program_id);
+
+        let transfer_to_taker_accounts = Transfer {
+            from: ctx
+                .accounts
+                .pda_deposit_token_account
+                .to_account_info()
+                .clone(),
+            to: ctx
+                .accounts
+                .taker_receive_token_account
+                .to_account_info()
+                .clone(),
+            authority: ctx.accounts.pda_account.clone(),
+        };
+
+        let seeds = &[&b"escrow"[..], &[bump_seed]];
+        let signer = &[&seeds[..]];
+
+        let transfer_to_taker_context = CpiContext::new_with_signer(
```

Comment:
One thing I started doing is constructing all `CpiContext` objects via `From` impls placed towards the bottom of the file in its own section.

```rust
impl From<&Exchange> for CpiContext<'_, '_, '_, 'info, Exchange<'info>> {
   fn from (accs: &Exchange) -> Self {
     ...
   }
}
```

It makes the program logic much more clear and removes all the CPI boilerplate out of the core program logic.

Then to do cpi one would do

```rust
        token::transfer(
            (&*ctx.accounts).into(),
            ctx.accounts.escrow_account.initializer_amount,
        )?;
```

Reply:
Hey @armaniferrante, I started off with this approach since it is a lot nicer, but it seems like you can't implement a trait twice for the same struct. (I'm pretty new to rust so lmk if wrong)

I have two different CpiContexts for the `token::transfer` in my `exchange` function, so I wanted to keep it consistent across the code. Is there a nice way to do this?

Reply:
I'd just implement regular methods instead of the trait.

```rust
impl Exchange {
   fn into_<my_cpi_method>(&self) -> CpiContext<...>
}
```

Reply:
Cool, that looks much cleaner now.